### PR TITLE
remove sensitive data from logs

### LINKED ISSUE #353: Redact sensitive fields in logs
## Expected Behavior
Logs should redact sensitive fields e.g: passwords or private keys. 

## Current Behavior
Logs tend to print everything in a provided message. For example: `http://localhost:8001/api/v1/destination_implementations/update` tends to print all of the credentials provided to the endpoint. 

## Steps to Reproduce
1. Create a destination in the UI
1. In the destination view in the UI, update the credentials for the destination
1. Save
1. Look at server logs

### LINKED ISSUE #1715: hide secrets in server logs for request / response bodies
## Tell us about the problem you're trying to solve
* we log the fully request and response for all API calls. some of the configuration objects contain secrets.

## Describe the solution you’d like
* prefer that we just hide fields with the `secrets` flag set. and that we can set an env variable to show them if we are debugging something in dev
* if for some reason that isn't doable simply not logging the `configuration` objects should be sufficient.

### PR DESCRIPTION
**Pull request by jrhizor**
closes https://github.com/airbytehq/airbyte/issues/1715
closes https://github.com/airbytehq/airbyte/issues/353

1. Add a patched version of the Java `LogsApi` client that can read `text/plain (the codegen version throws an exception)
2. Make all threads within the server actually log to the files (only the main thread logs would work before)
3. Make all threads within the scheduler actually log to the files (only the main thread logs would work before)
4. Return a 404 when a `NotFoundException` occurs. (before this would throw a 500 because of the exception even though it was actually a 404)
5. Create a custom logger that simplifies the request log output and makes it easier to modify. This logger doesn't log the responses and optionally logs the request body. There's an exclusion list that makes it easy to exclude typed requests that contain sensitive information. Unfortunately, it isn't easy to know what fields contain sensitive information because that data is stored in the spec, not the request made that conforms to the spec. I didn't want to force spec retrievals just for logging, which is why I'm just using exclusion lists.
   - Please let me know if you think I should be doing spec retrievals.
   - Please let me know if you think I should just exclude specific fields from the request types
   - Please let me know if you want me to add back in response logging. At the very least I'll keep the logs output removed because it makes the logs impossible to read. 
   - All of the above should be much easier to change now that we aren't using `LoggingFeature`
6. Test that one of the request types is hidden from the logs in the acceptance tests.

### CODE REVIEW

**Comment by jrhizor** [comment]
Example output:
```
2021-01-23 00:29:54 INFO  REQ 192.168.208.1 POST 204 /api/v1/destinations/delete - (REDACTED) Request body contains sensitive information!
2021-01-23 00:29:54 INFO  REQ 192.168.208.1 POST 200 /api/v1/source_definitions/list
2021-01-23 00:29:54 DEBUG cache hit: airbyte/source-postgres:0.1.10
2021-01-23 00:29:54 INFO  REQ 192.168.208.1 POST 200 /api/v1/sources/create - (REDACTED) Request body contains sensitive information!
2021-01-23 00:29:54 INFO  REQ 192.168.208.1 POST 200 /api/v1/destination_definitions/list
2021-01-23 00:29:54 DEBUG cache hit: airbyte/destination-postgres:0.1.13
2021-01-23 00:29:54 INFO  REQ 192.168.208.1 POST 200 /api/v1/destinations/create - (REDACTED) Request body contains sensitive information!
2021-01-23 00:29:54 INFO  enqueuing pending job for scope: 8731e783-be9d-4f5e-bb40-c5029c9e43b9
2021-01-23 00:29:54 INFO  Waiting for job id: 14
2021-01-23 00:30:02 INFO  REQ 192.168.208.1 POST 200 /api/v1/sources/discover_schema - (REDACTED) Request body contains sensitive information!
2021-01-23 00:30:02 INFO  REQ 192.168.208.1 POST 200 /api/v1/connections/create - {"name":"test-connection","sourceId":"8731e783-be9d-4f5e-bb40-c5029c9e43b9","destinationId":"2c7e9e16-2345-4af2-93ef-fa65220a4b2c","syncMode":"full_refresh","syncSchema":{"streams":[{"name":"public.id_and_name","cleanedName":"public_id_and_name","fields":[{"name":"id","cleanedName":"id","dataType":"number","selected":true},{"name":"name","cleanedName":"name","dataType":"string","selected":true}],"supportedSyncModes":["full_refresh","incremental"],"defaultCursorField":[],"cursorField":[]}]},"schedule":{"units":1,"timeUnit":"minutes"},"status":"active"}
2021-01-23 00:30:32 INFO  REQ 192.168.208.1 POST 200 /api/v1/logs/get - {"log_type":"server"}
2021-01-23 00:30:32 DEBUG cache hit: airbyte/source-postgres:0.1.10
2021-01-23 00:30:32 INFO  REQ 192.168.208.1 POST 204 /api/v1/sources/delete - (REDACTED) Request body contains sensitive information!
2021-01-23 00:30:32 INFO  REQ 192.168.208.1 POST 200 /api/v1/connections/get - {"connectionId":"d57edeab-6dd9-47a5-9fec-9c11aa0f3c11"}
2021-01-23 00:30:32 INFO  REQ 192.168.208.1 POST 200 /api/v1/connections/update - {"connectionId":"d57edeab-6dd9-47a5-9fec-9c11aa0f3c11","syncSchema":{"streams":[{"name":"public.id_and_name","cleanedName":"public_id_and_name","fields":[{"name":"id","cleanedName":"id","dataType":"number","selected":true},{"name":"name","cleanedName":"name","dataType":"string","selected":true}],"supportedSyncModes":["full_refresh","incremental"],"defaultCursorField":[],"cursorField":[]}]},"schedule":{"units":1,"timeUnit":"minutes"},"status":"deprecated"}
2021-01-23 00:30:32 DEBUG cache hit: airbyte/destination-postgres:0.1.13
2021-01-23 00:30:32 INFO  REQ 192.168.208.1 POST 204 /api/v1/destinations/delete - (REDACTED) Request body contains sensitive information!
2021-01-23 00:30:33 INFO  REQ 192.168.208.1 POST 200 /api/v1/logs/get - {"log_type":"server"}
```

File: airbyte-server/src/main/java/io/airbyte/server/RequestLogger.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import javax.ws.rs.container.ContainerResponseFilter;
+import javax.ws.rs.core.Context;
+import org.apache.commons.io.IOUtils;
+import org.eclipse.jetty.http.HttpStatus;
+import org.glassfish.jersey.message.MessageUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+// todo: make sure handler code has MDC functional
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/RequestLogger.java:55]:
```suggestion
```

File: airbyte-server/src/main/java/io/airbyte/server/RequestLogger.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      DestinationUpdate.class,
+      DestinationRecreate.class);
+
+  private static String redactSensitiveInfo(String requestBody) {
+    Optional<JsonNode> jsonNode = Jsons.tryDeserialize(requestBody);
+
+    if (jsonNode.isPresent()) {
+      for (Class<?> sensitiveClass : SENSITIVE_INFO_REQUESTS) {
+        Optional<?> object = Jsons.tryObject(jsonNode.get(), sensitiveClass);
+        if (object.isPresent()) {
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/RequestLogger.java:132]: is it always the `configuration` field that contains the sensitive info? if so i might argue that it would be nice to print everything but that field so we can have a sanity check that the rest looks good. but if it's not dead simple to do, ignore.

File: airbyte-api/src/main/java/io/airbyte/api/client/PatchedLogsApi.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.function.Consumer;
+import org.apache.commons.io.FileUtils;
+
+/**
+ * This class is a copy of {@link LogsApi} except it allows Accept: text/plain. Without this
```
Comment (cgardens) [inline airbyte-api/src/main/java/io/airbyte/api/client/PatchedLogsApi.java:44]: vom.

**Review by cgardens — APPROVED** [review]
thanks for doing this. i had no idea that this task was going to be so obnoxious.


UPDATED


Please let me know if you think I should be doing spec retrievals. **no**
Please let me know if you think I should just exclude specific fields from the request types **mentioned below but if it's always the configuration field i might just do that if it's not too difficult**
Please let me know if you want me to add back in response logging. At the very least I'll keep the logs output removed because it makes the logs impossible to read. **i'm fine with removing for now. we can bring it back if missing it.**

in short. i like what you did. if we can just redact a specific field let's do it. otherwise i'm happy with it as is.

**Comment by cgardens** [comment]
updated my review to answer the posed questions.

File: airbyte-server/src/main/java/io/airbyte/server/RequestLogger.java (hunk 2 above)
Comment (jrhizor) [reply airbyte-server/src/main/java/io/airbyte/server/RequestLogger.java:132]: Yeah it looks like `connectionConfiguration` only right now. I'm going to remove it, but only from the top level for now.

**Comment by jrhizor** [comment]
Updated and now logs look like:
```
2021-01-25 17:52:27 INFO  REQ 172.18.0.1 POST 200 /api/v1/destinations/create - {"destinationDefinitionId":"25c5221d-dce2-4163-ade9-739ef790f503","connectionConfiguration":"REDACTED","workspaceId":"5ae6b09b-fdec-41af-aaf7-7d94cfc33ef6","name":"AccTestDestination-0812ac56-9803-4362-93dd-7a0469206fa0"}
```

**Comment by cgardens** [comment]
good. i like it.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/SchedulerApp.java [hunk 4]
```diff
@@ -94,12 +95,15 @@ public void start() throws IOException {
     final JobScheduler jobScheduler = new JobScheduler(jobPersistence, configRepository);
     final JobSubmitter jobSubmitter = new JobSubmitter(workerThreadPool, jobPersistence, configRepository, workerRunFactory);
 
+    Map<String, String> mdc = MDC.getCopyOfContextMap();
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/SchedulerApp.java:98]: I assume this change to enable logging in threads?

I am surprised that you have to do anything for the threads to log properly. If it doesn't work out of the box, it means that our log4j configuration file isn't properly configured.

Comment (michel-tricot) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/SchedulerApp.java:98]: How did you decide it was the right solution to fix the pb?
